### PR TITLE
Fix 14561 - Large enums cannot be parsed due to too many recursive template expansions

### PR DESCRIPTION
**Pull request by Biotronic**
As discussed in https://issues.dlang.org/show_bug.cgi?id=14561, this pull makes EnumMembers use divide-and-conquer (through staticMap), and adds to NoDuplicates support for lists longer than 500 elements.

### CODE REVIEW

**Comment by yebblies** [comment]
Fails on autotester, and indentation needs fixing.

**Comment by McSherry** [comment]
PR title should probably be expanded: http://wiki.dlang.org/Pull_Requests#Create_a_pull_request

> Choose a title for your pull request that clearly states what it does. When fixing a bug, the usual thing to do is to use the summary from the bugzilla report. Eg a title like "Fix 3797" or "Issue 3797" contains much less information than "Fix Issue 3797 - Regression(2.038): Implicit conversion between incompatible function pointers" and requires a lot more effort for the reviewers to determine if it is something they are interested in.

**Comment by schveiguy** [comment]
> PR title should probably be expanded

Yes, thanks for noticing. That's a pet peeve of mine.

**Comment by quickfur** [comment]
ping
There are merge conflicts, please rebase.

**Comment by Hackerpilot** [comment]
@Biotronic can you update this pull so that it applies cleanly?

**Comment by quickfur** [comment]
ping @Biotronic 
Rebase please?

**Comment by Biotronic** [comment]
I'm back, baby! Sorry about the wait, I've had the last few months of my life ruined by near-constant migraines, but have finally found medication that works.

**Comment by quickfur** [comment]
Good to hear!  Could you also squash the commits?

**Comment by Biotronic** [comment]
Is that good enough, or should I go all the way back to f9e0862?

**Comment by quickfur** [comment]
What is 0e3b7ff for? It looks a bit fishy to me. Technically we're supposed to be working on topic branches so that we can merge back to master here as a single branch.  But no matter, I don't think it's a big deal.

LGTM.

**Comment by Biotronic** [comment]
It's when I pulled from https://github.com/D-Programming-Language/phobos/. Maybe I'm doing it wrong?

File: std/traits.d [hunk 1]
```diff
@@ -3467,6 +3474,45 @@ unittest
     static assert(__traits(identifier, EnumMembers!E[2]) == "b");
 }
 
+unittest { // Huge enums
+    enum TLAs { aa, ab, ac, ad, ae, af, ag, ah, ai, aj, ak, al, am, an, ao, ap, aq, ar, as, at,
```
Comment (BBasile) [inline std/traits.d]: Wouldn't it be acceptable to use template meta programming to generate this ?!

```D
string genEnum()
{
    string result = "enum TLAs {";
    foreach(c0; 'a'..'z')
        foreach(c1; 'a'..'z')
    {
        result ~= '_';
        result ~= c0;
        result ~= c1;
        result ~= ',';
    }
    result ~= '}';
    return result;
}
mixin(genEnum);
```

File: std/meta.d [hunk 2]
```diff
@@ -355,6 +376,62 @@ unittest
...
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int,
+        int,int,int,int,int,int,int,int,int,int);
```
Comment (quickfur) [inline std/meta.d]: This can be replaced with a much more concise version by using `std.meta.Repeat`.

**Comment by Biotronic** [comment]
Cleaned up commits and unit test code. I seem to be getting the hang of this Git thing now.

File: std/traits.d [hunk 3]
```diff
@@ -3459,6 +3466,31 @@ unittest    // duplicated values
     static assert([ EnumMembers!A ] == [ A.a, A.b, A.c, A.d, A.e ]);
 }
 
+unittest // huge enums
```
Comment (BBasile) [inline std/traits.d]: Could you comment with `\\ bugzilla 14561` instead ? It seems that this is widely adopted.
If someone read the code, with  `\\ huge enums`, it's not clear that it used to be a bug.

Comment (ZombineDev) [inline std/traits.d]: @BBasile you mean `// bugzilla 14561`, right?

Comment (BBasile) [inline std/traits.d]: Yes.

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by BBasile** [comment]
Yes plz, this is a bug fix.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by Laeeth** [comment]
Did this PR just raise the limit?  It sounds like it was intended to work with any size enums, but on a large enum of about 20,000 I get recursive template expansion errors still.

**Comment by BBasile** [comment]
It's not yet in phobos, latest release was a dot release so it just included the regressions.

**Comment by JackStouffer** [comment]
@Laeeth if you need this commit, please check out the nightly releases: https://dlang.org/download.html#dmd-nightly

**Comment by Biotronic** [comment]
@Laeeth: Yes, it only raises the limit. It changes the implementation to a divide-and-conquer pattern, so the limit is 2^500 instead of 500. You're likely to run into other problems at that point.

**Comment by Laeeth** [comment]
Thanks, for the colour everyone.   I switched to using a file loaded at launch instead.

**Pull request merged**